2015/presentations/Expert system in Python in 5 minutes/text.md

```diff
@@ -0,0 +1,144 @@
+
+    def dumpOutput(self):
+        result = clips.TraceStream.Read()
+        if result:
+            self.log.info('CLIPS traceback:\n%s' % result)
+
+        result = clips.StdoutStream.Read()
+        if result:
+            self.log.info('CLIPS stdout:\n%s' % result)
+
+        result = clips.ErrorStream.Read()
+        if result:
+            self.log.warning('Errors:\n%s' % result)
+
+
+## Facts
+
+Before we are going to dive into details how to process facts and take some actions what first comes first - loading facts. To simplify this article let's assume that fact is a very simple state which is going to be checked by rules engine. If some states of an entity are going to match any of the rules then that particular rule(s) is going to be fired. Fired means that engine can modify any of its internal objects (another facts) or for example call one of the registered before Python functions.
+
+How to load fact into rules engine from Python, example:
+
+    instance = clips.FindInstance(some_object_hash)
+    
+Now modify facts
+
+    instance.Slots['some_attribute'] = serializable_python_variable
+    
+We can also assign a fact which is a reference to another instance
+
+    instance.Slots['some_attribute'] = clips.FindInstance(another_object_hash)
+    
+We can also send singnal to the engine about some changes of facts, so engine is going to be aware of our assigning process.
+
+    clips.Eval('(modify-instance %s (some_fact (create$)))' % instance.Name)
+    
+    
+## Let the fire burn
+
+Once you know how to register Python functions and assign facts it is time to see how exactly you are going to use rules engine. Again to simplify the description I will assume that you alredy assigned facts, loaded rules and python functions so let's focus only on the execution of the rules.
+
+Below function can be started as a dedicated thread which has to go into ininity loop so we are going to keep processing rules over and over again (based on changeable facts)
+
+    def reactor():
+        self.log.info('Start processing')
+        rules_fired = clips.Run(5000) # max rules to fire for that loop
+        self.log.info('CLIPS fired %s rules' % (rules_fired))
+        # here you can dump output to debug what was happening
+
+Now as for debug. It is very important to notice that if debug is actually enabled the whole engine is going to be few times slowe, which of course is not recommended in production use. Another thing...it leaks realy badly with debug being on.
+
+
+## Final note
+
+Of course I am fully aware of the fact that pyCLIPS project is not being updated since 2008. This is not the main point in that article. I just wanted to show how to use C module for Python with external laguage and its semi-engine (something like LISP) which allows you to write much simpler and cleaner code for expert systems.
```

Comment:
Let me put some more light in introduction. I will explain more. It is worth of using pyCLIPS in production if you don't plan to build very complex systems. Yes, with debug being on pyCLIPS is really slow and it leaks pretty badly, so debug is recommended as a last resort.